pkg/controlplane/deleted_kinds.go

```diff
@@ -0,0 +1,167 @@
+		minorString := strings.Split(currentVersion.Minor, "+")[0]
+		minorString = strings.Split(minorString, "-")[0]
+		minorString = strings.Split(minorString, ".")[0]
+		currentMinor64, err := strconv.ParseInt(minorString, 10, 32)
+		if err != nil {
+			return nil, err
+		}
+		ret.currentMinor = int(currentMinor64)
+	}
+
+	ret.isAlpha = strings.Contains(currentVersion.GitVersion, "alpha")
+
+	if envString, ok := os.LookupEnv("KUBE_APISERVER_STRICT_REMOVED_API_HANDLING_IN_ALPHA"); !ok {
+		// do nothing
+	} else if envBool, err := strconv.ParseBool(envString); err != nil {
+		return nil, err
+	} else {
+		ret.strictRemovedHandlingInAlpha = envBool
+	}
+	if envString, ok := os.LookupEnv("KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE"); !ok {
+		// do nothing
+	} else if envBool, err := strconv.ParseBool(envString); err != nil {
+		return nil, err
+	} else {
+		ret.serveRemoveAPIsOneMoreRelease = envBool
+	}
+
+	return ret, nil
+}
+
+func (e *resourceExpirationEvaluator) shouldServe(resourceServingInfo rest.Storage) bool {
+	versionedPtr := resourceServingInfo.New()
+	removed, ok := versionedPtr.(removedInterface)
+	if !ok {
+		return true
+	}
+	majorRemoved, minorRemoved := removed.APILifecycleRemoved()
+	if e.currentMajor < majorRemoved {
+		return true
+	}
+	if e.currentMajor > majorRemoved {
+		return false
+	}
+	if e.currentMinor < minorRemoved {
+		return true
+	}
+	if e.currentMinor > minorRemoved {
+		return false
+	}
+	// at this point major and minor are equal, so this API should be removed when the current release GAs.
+	// If this is an alpha tag, don't remove by default, but allow the option.
+	// If the cluster-admin has requested serving one more release, allow it.
+	if e.isAlpha && e.strictRemovedHandlingInAlpha { // don't serve in alpha if we want strict handling
+		return false
+	}
+	if e.isAlpha { // alphas are allowed to continue serving expired betas while we clean up the test
+		return true
+	}
+	if e.serveRemoveAPIsOneMoreRelease { // cluster-admins are allowed to kick the can one release down the road
+		return true
+	}
+	return false
+}
+
+type removedInterface interface {
+	APILifecycleRemoved() (major, minor int)
+}
+
+// removeDeletedKinds inspects the storage map and modifies it in place by removing storage for kinds that have been deleted.
+// versionedResourcesStorageMap mirrors the field on APIGroupInfo, it's a map from version to resource to the storage.
+func (e *resourceExpirationEvaluator) removeDeletedKinds(groupName string, versionedResourcesStorageMap map[string]map[string]rest.Storage) {
+	versionsToRemove := sets.NewString()
+	for apiVersion, versionToResource := range versionedResourcesStorageMap {
+		resourcesToRemove := sets.NewString()
+		for resourceName, resourceServingInfo := range versionToResource {
+			if !e.shouldServe(resourceServingInfo) {
+				resourcesToRemove.Insert(resourceName)
+			}
+		}
+
+		for resourceName := range versionedResourcesStorageMap[apiVersion] {
```

Comment:
why this double loop, and `versionedResourcesStorageMap[apiVersion]` vs. `versionToResource`?

Reply:
> why this double loop, and `versionedResourcesStorageMap[apiVersion]` vs. `versionToResource`?

10 lines up, the godoc describes the args.  This mirrors the structure in `APIGroupInfo`, so per the comment `it's a map from version to resource to the storage.`.  If I change it here, the usage becomes more difficult and the results cannot be directly used.